lib/highline.rb

```diff
@@ -677,20 +449,17 @@ def indentation
   def indent(increase=1, statement=nil, multiline=nil)
     @indent_level += increase
     multi = @multi_indent
-    @multi_indent = multiline unless multiline.nil?
+    @multi_indent = multiline || @multi_indent
```

Comment:
@abinoam You can go ahead and do it yourself. =)

I would like to give you a better feedback but i'm not used to the issues this code solve neither the code itself so anything i say here would not be helpful or precise.
Give me some time to go over the lib and understand what it does and the design decisions behind it. I want to see the big picture to be able to help you as much as possible here.
Overall the changes are awesome, you did a lot of encapsulation and good refactoring just as @apotonick mentioned.